### ISSUE TITLE
Why not return an empty realmResults when using "in" query with empty values instead of throwing an exception?

### ISSUE THREAD

**Issue by gpulido** [realm/realm-java#4011]
#### Goal

Obtain a realmResults using the "in" query operator on where the values list is empty.
Currently this throws an exception that has to be handled.

#### Expected Results

In set theory, the empty set already is comply by the full set or the empty set, so instead of throwing an exception return an empty realmresults

#### Actual Results
An exception

**Comment by Zhuinden** [comment]
Well yes, except there's [no such thing as an empty results](https://github.com/realm/realm-java/issues/1862), unless you create query conditions that can never be true.

**Comment by gpulido** [comment]
I agree but the in operator is the same as an "equalto" as it filters the collection. So why the equalto can return an "empty" realmresults and the "in" operator can't?

**Comment by Zhuinden** [comment]
Technically I'd assume that providing `in(null)` or `in(Set<?> emptySet)` would do a no-op, but I guess that's not the case, it does throw an exception.

Either way, an empty results wouldn't make sense. But a no-op would make sense.

**Comment by gpulido** [comment]
I dont' think that the problem is the empty results as they can be obtained using an equalto that doesn't return any match, the problem is that when you have the empty list to concatenate to you can't generate the query. However from the set theory point of view, returning an empty collection or the entire collection would be a better aproach that just throwing an exception.

I guess I will keep doing what I'm doing: checking if the list is empty before and if it is empty just generate a criteria that could not be satisfied.
Thanks for the support.

**Comment by kneth** [comment]
@gpulido Initially we have been thinking that if you provide an empty list for `in`, it should indicate a programming error. I'll add this issue to our backlog so we remember to discuss it at our next revision of the API.

**Comment by gpulido** [comment]
@kneth Maybe is my use case the one that can be resolved using other mechanism:
My model is the following:
```Java
ParentObject extends RealmObject
{
 @PrimaryKey
    private String UUID;
    RealmList<ChildObject> childObjects;
}

ChildObject extends RealmObject
{
 @PrimaryKey
    private String UUID;
    boolean isCompleted
}
```
I need to pick the first childObject from each ParentObject that isCompleted = false, and create a RealmResults<ChildObject> that feeds a RecycledView (for example)
So I came with this solution (I'm using retrolambda and annimon streams)
```Java
 RealmResults<ParentObject > parents= realm.where(Contender.class) .findAll();

childObjectsUUID = Stream.of(parents).map(parents-> parents.getChildObjects().where().equalTo("isCompleted", false).findFirst()).map(ChildObject ::getUUID)
                    .toArray(String[]::new);

//Finally obtain the realmResults that I need:
realm.where(ChildObject .class).in("UUID", childObjectsUUID ).findAll();
```
The "childObjectsUUID" sometimes is empty and hence this issue :)
This has been done this way as I don't have a find way to flatten the hierarchy from realm.
Maybe there is a better way to do this.

**Comment by cmelchior** [comment]
Would also be interesting how other Databases/libraries handle this scenario

**Comment by cmelchior** [comment]
We actually had this discussion:  https://github.com/realm/realm-java/pull/3133#r73102437
Conclusion seems that we should support it.

**Comment by isoiphone** [comment]
Any ETA on this? The Android API is different here than iOS. Feels more like a bug than a feature/improvement.

**Comment by cmelchior** [comment]
Fixed in #5898
